### PR TITLE
Allow Different Compute Layout for Attention

### PR DESCRIPTION
**Pull request by morgandu**
# Checklist
This PR introduced compute layout control to allowed a different compute layout for attention
- [x] Attention Unit Tests for different compute layout
- [x] Microbenchmark - Performance
- [x] E2E Serving - Accuracy and Performance

# Setup

- v5e8
- llama2-7b
- quant_mode
    - w-bf16-kv-bf16
    - for w-i8-kv-i8 see #708
- model_mode
    - base for performance
    - chat for accuracy

# Results and Analysis

The goal of introducing the new compute layout is to potentially avoid cache layout tuning, though we still can tune the cache layout to seek and verify for the best performance. 

### Annotation
- b: batch
- t: query_length
- h: query_heads
- d: kv_dimension
- s: kv_length
- k: kv_heads

### Layout
- 0123 | bthd | bskd
- 0213 | bhtd | bksd
- 1203 | thbd | skbd

## Summary

Existing attention compute layout is 0123, and we introduced a different compute layout 0213, which is of a layout that's TPU friendly. 

We introduced 0213 compute layout to verify:
    - if and how much 0213 has direct impact on performance on the default cache layout, i.e. same layout as compute layout
    - if and how much different compute layout and different cache layouts have a composite impact on performance

##### Cache layout 1203-1203
With the existing cache layout was 1203-1203, with throughput 2591.642232 tokens/s, this was improved from the default cache layout 0123-0123 about 3x. 

##### Cache layout 2013-2013
After layout tuning, we got optimal prefill-ar cache layout as 2013-2013, with throughput 3347.180221 tokens/s, which was 29% improvement.

##### Cache layout 0213-0213
With the two cache in the same layout as compute, i.e. 0213-0213 (xprof: https://xprof.corp.google.com/overview_page/morgandu-12159058496322304249), we got 3273.96 tokens/s, this is of the top performance after we verified with layout tuning.

##### Cache layout 0213-0132
The tuned cache layout that give us the best throughput 3329.45 tokens/s is 0213-0132 (xprof: https://xprof.corp.google.com/overview_page/morgandu-5743582688063478644)

### Accuracy

No regression on Rouge scores between 0123 and 0213

```
{'rouge1': 42.1738, 'rouge2': 19.6973, 'rougeL': 26.9088, 'rougeLsum': 39.6794, 'gen_len': 1144204, 'gen_num': 995}
```

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -31,6 +31,40 @@
...
+        "base_output_directory=gs://test-maxtext-output",
+        "prefill_cache_axis_order=0,2,1,3",
+        "ar_cache_axis_order=0,2,1,3",
+        "compute_axis_order=0,2,1,3",
+        "reshape_q=true",
+        "per_device_batch_size=24",
+        "quantization=int8",
+        "quantize_kvcache=True",
+      ]
+    },
```
Comment (vipannalla) [inline .vscode/launch.json:68]: n00b question -- Is this to auto run for every local commit/amend on vscode? or just for convinience?

File: MaxText/pyconfig.py [hunk 2]
```diff
@@ -52,6 +52,11 @@ def string_to_bool(s: str) -> bool:
 _yaml_types_to_parser = {str: str, int: int, float: float, bool: string_to_bool}
 
 
+def validate_compute_axis_order(s: str) -> None:
+  valid_compute_axis_order = ("0,1,2,3", "0,2,1,3")
```
Comment (vipannalla) [inline MaxText/pyconfig.py:56]: Why not allow other layouts? Does the code break with others or just run slow? I'd allow other layouts similar to prefill_cache_axis_order/ar_cache_axis_order. 
Maybe remove the exception and just print a warning that others are untested?

File: MaxText/tests/attention_test.py [hunk 3]
```diff
@@ -257,71 +258,73 @@ def tpu_kernel_attention_helper(self, num_kv_heads):
     )
 
   @pytest.mark.tpu
-  def test_dot_product_1203_1203(self):
-    self.dot_product_attention_helper(
-      prefill_cache_axis_order=(1,2,0,3),
-      ar_cache_axis_order=(1,2,0,3)
-    )
+  def test_dot_product_cache_axis_order(self):
```
Comment (vipannalla) [inline MaxText/tests/attention_test.py:261]: Thanks for adding unitests!

File: MaxText/layers/attentions.py [hunk 4]
```diff
@@ -68,6 +68,12 @@
 # pytype: disable=attribute-error
 
 
+def validate_compute_axis_order(s: str) -> None:
```
Comment (vipannalla) [inline MaxText/layers/attentions.py:71]: duplicate, remove. I see this method in pyconfig.py as well, which is right place to validate at the beginning of the run.

File: MaxText/pyconfig.py (hunk 2 above)
Comment (vipannalla) [reply MaxText/pyconfig.py:56]: nvm, I noticed in attention.py you specifically look for those two.

File: MaxText/layers/attentions.py [hunk 5]
```diff
@@ -359,13 +366,21 @@ def qk_product(self, query: Array, key: Array, q_seq_len: int) -> Array:
...
+      if self.reshape_q and q_seq_len == 1:
+        query = jnp.broadcast_to(query, (b, 2, n_kv, n // n_kv, d))
+      result = jnp.einsum("btkgd,bskd->bkgts", query, key)
+    elif self.compute_axis_order == (0,2,1,3):
+      query = jnp.transpose(query, axes=self.compute_axis_order)
+      key = jnp.transpose(key, axes=self.compute_axis_order)
+      query = jnp.reshape(query, (b, n_kv, n // n_kv, t, d))
+      if self.reshape_q and q_seq_len == 1:
+        query = jnp.broadcast_to(query, (b, n_kv, n // n_kv, 2, d))
+      result = jnp.einsum("bkgtd,bksd->bkgts", query, key)
```
Comment (vipannalla) [inline MaxText/layers/attentions.py:380]: This lgtm for now. Since we already know 0 = b, 1 = t, 2= n etc.., can this code be made generic to support all layouts?

**Review by vipannalla — COMMENTED** [review]
Looks good to me

File: .vscode/launch.json (hunk 1 above)
Comment (morgandu) [reply .vscode/launch.json:68]: no, I added this when I was debugging on my local vscode, which is helpful for other engineers, but you may need to change the flags depends on your run.

File: MaxText/tests/attention_test.py (hunk 3 above)
Comment (morgandu) [reply MaxText/tests/attention_test.py:261]: of course!

File: MaxText/layers/attentions.py (hunk 4 above)
Comment (morgandu) [reply MaxText/layers/attentions.py:71]: Actually this is a different validate. Since the base.yml only allows string, so the pyconfig validates the flag's string value. In attention module, the axis_orders are really used as tuple, and it can be hard code / overwrite in different model class initiation, i.e. https://github.com/google/maxtext/blob/main/MaxText/layers/gpt3.py#L229

I don't want to risk with possible hard coded axis_orders passing from somewhere other than yaml flags.

File: MaxText/layers/attentions.py (hunk 5 above)
Comment (morgandu) [reply MaxText/layers/attentions.py:380]: A lot of other places needs to be changed to support more layouts, I have experienced some loud and silent bug when I was trying to add one more layout till I gave up.

Also, I don't think it's necessary to support all layouts. Since it was recommended that bh** is one of the friendly layout. I'd say if we end up really need to include new layouts then let's revisit it!

File: MaxText/configs/base.yml [hunk 6]
```diff
@@ -312,10 +312,13 @@ inference_microbenchmark_log_file_path: ""
...
-prefill_key_axis_order: "1,2,0,3"
-prefill_value_axis_order: "1,2,0,3"
-ar_key_axis_order: "1,2,0,3"
-ar_value_axis_order: "1,2,0,3"
+prefill_cache_axis_order: "1,2,0,3"
+ar_cache_axis_order: "1,2,0,3"
+
+# Compute layout control
+# Default layout: 0,1,2,3 ; BATCH, LENGTH, HEAD, D_KV
+compute_axis_order: "0,1,2,3"
```
Comment (gobbleturk) [inline MaxText/configs/base.yml:320]: Can we add a comment here that the only supported ones are (0,1,23) and (0,2,1,3)?

**Review by gobbleturk — APPROVED** [review]
